Antidote/VideoAndPreviewView.m

```diff
@@ -0,0 +1,139 @@
+//
+//  VideoAndPreviewView.m
+//  Antidote
+//
+//  Created by Chuong Vu on 8/11/15.
+//  Copyright (c) 2015 dvor. All rights reserved.
+//
+
+#import "VideoAndPreviewView.h"
+#import "Masonry.h"
+
+static const CGFloat kPreviewViewWidth = 75.0;
+static const CGFloat kPreviewViewHeight = 100.0;
+@interface VideoAndPreviewView ()
+
+@property (strong, nonatomic) UIView *previewView;
+
+@end
+
+@implementation VideoAndPreviewView
+
+#pragma mark Life cycle
+
+- (instancetype)init
+{
+    self = [super init];
+
+    if (! self) {
+        return nil;
+    }
+
+    self.backgroundColor = [UIColor blackColor];
+
+    [self createPreviewView];
+
+    [self installConstraints];
+
+    return self;
+}
+
+#pragma mark - View setup
+
+- (void)createPreviewView
+{
+    self.previewView = [UIView new];
+    self.previewView.backgroundColor = [UIColor blackColor];
+    self.previewView.hidden = YES;
+
+    [self addSubview:self.previewView];
+}
+
+- (void)layoutSubviews
+{
+    [super layoutSubviews];
+    [self layoutIfNeeded]; // is this needed?
+    [self adjustPreviewLayer];
+
+}
+- (void)installConstraints
+{
+    [self.previewView makeConstraints:^(MASConstraintMaker *make) {
+        make.bottom.equalTo(self);
+        make.right.equalTo(self);
+        make.height.equalTo(kPreviewViewHeight);
+        make.width.equalTo(kPreviewViewWidth);
+    }];
+}
+
+#pragma mark - Public
+
+- (void)setVideoView:(UIView *)videoView
+{
+    if (videoView == _videoView) {
+        return;
+    }
+
+    if (_videoView) {
+        [_videoView removeFromSuperview];
+    }
+
+    _videoView = videoView;
+
+    if (! _videoView) {
+        return;
+    }
+
+    [self insertSubview:_videoView belowSubview:self.previewView];
+
+    [self.videoView makeConstraints:^(MASConstraintMaker *make) {
+        make.bottom.equalTo(self.previewView);
+        make.left.equalTo(self);
+        make.right.equalTo(self);
+        make.top.equalTo(self);
+    }];
+}
+
+- (void)setPreviewLayer:(CALayer *)previewLayer
+{
+    dispatch_async(dispatch_get_main_queue(), ^{
```

Comment:
Yea you are right. Though setting it to nil is not on the main queue though. I have to redo this logic, since even removing the layer from the superlayer while on the main queue causes issues/slowness.